File: android/app/src/main/java/com/apoorvdarshan/calorietracker/models/UserExercise.kt
package com.apoorvdarshan.calorietracker.models

import com.apoorvdarshan.calorietracker.data.ExerciseItem
import java.util.UUID

object UserExercise {
    const val ID_PREFIX = "user_exercise_"

    fun newId(): String = "$ID_PREFIX${UUID.randomUUID()}"

    fun isUserExercise(id: String): Boolean = id.startsWith(ID_PREFIX)

    fun photoFilename(exerciseId: String): String {
        val uuid = exerciseId.removePrefix(ID_PREFIX)
        return "exercise_$uuid.jpg"
    }

    /** Fresh on-disk name for a new/replaced custom exercise photo. */
    fun newPhotoFilename(): String = "exercise_${UUID.randomUUID()}.jpg"

    fun isUserPhotoFilename(filename: String): Boolean =
        filename.startsWith("exercise_") && filename.endsWith(".jpg")
}

data class UserExerciseDraft(
    val name: String = "",
    val instructions: String = "",
    val level: String = "Beginner",
    val force: String = "Unspecified",
    val mechanic: String = "Unspecified",
    val category: String = "Strength",
    val equipment: String = "Unspecified",
    val primaryMuscles: List<String> = emptyList(),
    val secondaryMuscles: List<String> = emptyList(),
    val photoBytes: ByteArray? = null,
    val removePhoto: Boolean = false
) {
    val trimmedName: String get() = name.trim()

    val instructionLines: List<String>
        get() = instructions.lines().map { it.trim() }.filter { it.isNotEmpty() }

    fun toExerciseItem(id: String, imagePaths: List<String>): ExerciseItem = ExerciseItem(
        id = id,
        name = trimmedName,
        level = level,
        imagePaths = imagePaths,
        force = force,
        mechanic = mechanic,
        category = category,
        equipment = equipment,
        primaryMuscles = primaryMuscles,
        secondaryMuscles = secondaryMuscles,
        instructions = instructionLines
    )

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is UserExerciseDraft) return false
        return name == other.name &&
            instructions == other.instructions &&
            level == other.level &&
            force == other.force &&
            mechanic == other.mechanic &&
            category == other.category &&
            equipment == other.equipment &&
            primaryMuscles == other.primaryMuscles &&
            secondaryMuscles == other.secondaryMuscles &&
            removePhoto == other.removePhoto &&
            (photoBytes contentEquals other.photoBytes)
    }

    override fun hashCode(): Int {
        var result = name.hashCode()
        result = 31 * result + instructions.hashCode()
        result = 31 * result + level.hashCode()
        result = 31 * result + force.hashCode()
        result = 31 * result + mechanic.hashCode()
        result = 31 * result + category.hashCode()
        result = 31 * result + equipment.hashCode()
        result = 31 * result + primaryMuscles.hashCode()
        result = 31 * result + secondaryMuscles.hashCode()
        result = 31 * result + removePhoto.hashCode()
        result = 31 * result + (photoBytes?.contentHashCode() ?: 0)
        return result
    }
}

File: ios/calorietracker/Models/UserExercise.swift
import Foundation

enum UserExercise {
    static let idPrefix = "user_exercise_"

    static func newID() -> String { "\(idPrefix)\(UUID().uuidString)" }

    static func isUserExercise(_ id: String) -> Bool { id.hasPrefix(idPrefix) }

    static func photoFilename(forExerciseID id: String) -> String {
        let uuid = id.replacingOccurrences(of: idPrefix, with: "")
        return "exercise_\(uuid).jpg"
    }

    /// Fresh on-disk name for a new/replaced custom exercise photo.
    static func newPhotoFilename() -> String { "exercise_\(UUID().uuidString).jpg" }

    static func isUserPhotoFilename(_ filename: String) -> Bool {
        filename.hasPrefix("exercise_") && filename.hasSuffix(".jpg")
    }
}

struct UserExerciseDraft: Equatable {
    var name: String = ""
    var instructions: String = ""
    var rawLevel: String = "Beginner"
    var force: String = "Unspecified"
    var mechanic: String = "Unspecified"
    var category: String = "Strength"
    var rawEquipment: String = "Unspecified"
    var primaryMuscles: [String] = []
    var secondaryMuscles: [String] = []
    var photoData: Data?
    var removePhoto = false

    var trimmedName: String {
        name.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    var instructionLines: [String] {
        instructions
            .split(whereSeparator: \.isNewline)
            .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
            .filter { !$0.isEmpty }
    }

    func libraryItem(id: String, imagePaths: [String]) -> ExerciseLibraryItem {
        ExerciseLibraryItem(
            id: id,
            name: trimmedName,
            rawLevel: rawLevel,
            imagePaths: imagePaths,
            force: force,
            mechanic: mechanic,
            category: category,
            rawEquipment: rawEquipment,
            primaryMuscles: primaryMuscles,
            secondaryMuscles: secondaryMuscles,
            instructions: instructionLines
        )
    }
}
